### PR TITLE
refactor: split click UI from business logic for `data` cmds

### LINKED ISSUE #2402: Split `ilab data generate` click function apart from functionality
**What is the current behavior of feature you are looking to change?**
`ilab data generate` the click cmd is tied too tightly to the backend implementation

**What is the proposed behavior of feature you are looking to change?**
split the click cmd into a cli package and the functionality into a utils package

**What is the benefit of the design change you are proposing?**
this allows the CLI to follow a client server model for a future REST API and for multi processing execution

If applicable, how would backwards-compatability be handled?
N/A

Additional context

### PR DESCRIPTION
**Pull request by jaideepr97**
splits click UI entry points from business logic and places them in a new `instructlab/cli` package 
for data commands `generate` and `list`

Resolves #2402 

**Checklist:**

- [x] **Commit Message Formatting**: Commit titles and messages follow guidelines in the
  [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/#summary).
- [ ] [Changelog](https://github.com/instructlab/instructlab/blob/main/CHANGELOG.md) updated with breaking and/or notable changes for the next minor release.
- [ ] Documentation has been updated, if necessary.
- [ ] Unit tests have been added, if necessary.
- [ ] Functional tests have been added, if necessary.
- [ ] E2E Workflow tests have been added, if necessary.

### CODE REVIEW

File: src/instructlab/model/backends/vllm.py [hunk 1]
```diff
@@ -207,11 +210,6 @@ def format_template(model_family: str, model_path: pathlib.Path) -> str:
     return prefix + template
 
 
-def contains_argument(prefix: str, args: typing.Iterable[str]) -> bool:
```
Comment (danmcp) [inline src/instructlab/model/backends/vllm.py:210]: This is also used from serve and evaluate

File: src/instructlab/cli/data/list.py [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+    """lists datasets"""
+
+    data: List[List[Path]] = []
+
+    dirs = [Path(dir) for dir in dataset_dirs]
+
+    try:
+        data = list_data(dirs)
+    except OSError as exc:
+        click.secho(f"failed to list datasets with exception: {exc}")
```
Comment (danmcp) [inline src/instructlab/cli/data/list.py:39]: nit: Failed should be capitalized

**Review by danmcp — CHANGES_REQUESTED** [review]

File: src/instructlab/model/backends/vllm.py (hunk 1 above)
Comment (jaideepr97) [reply src/instructlab/model/backends/vllm.py:210]: good catch, thanks! updated

I'm surprised the import didn't break - seems like python may be able to handle transitive imports somehow?

File: src/instructlab/cli/data/list.py (hunk 2 above)
Comment (jaideepr97) [reply src/instructlab/cli/data/list.py]: fixed, thanks

**Review by alinaryan — APPROVED** [review]
Thanks JD!

**Review by alinaryan — COMMENTED** [review]
One question: are you planning to cover testing in this PR?

**Comment by jaideepr97** [comment]
> One question: are you planning to cover testing in this PR?

Good question
I think we may want to handle all the tests together - I think we will need to figure out how we want to structure the testing package, and where the tests themselves should live, where the common functions should be imported from etc

**Comment by nathan-weinberg** [comment]
@Mergifyio refresh

**Comment by nathan-weinberg** [comment]
@jaideepr97 please rebase this PR on top of `main` so we can run the large E2E job against it, thanks!

**Comment by danmcp** [comment]
@mergifyio rebase

File: src/instructlab/cli/data/data.py [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+    cls=clickext.LazyEntryPointGroup,
+    ep_group="instructlab.command.data",
+)
+@click.pass_context
+def data(ctx):
+    """Command Group for Interacting with the Data generated by InstructLab.
+
+    If this is your first time running ilab, it's best to start with `ilab config init` to create the environment.
+    """
+    ctx.obj = ctx.parent.obj
```
Comment (cdoern) [inline src/instructlab/cli/data/data.py:23]: can we do the ctx assigning after the storage check?

File: src/instructlab/data/list.py [hunk 4]
```diff
@@ -2,50 +2,40 @@
...
-def list_data(ctx, dataset_dirs: list[str]):
-    """lists datasets"""
     data: List[List[str]] = []
     for directory in dataset_dirs:
         dirpath = Path(directory)
         directories: List[Path] = [dirpath]
         top_level_dir = dirpath
         if not dirpath.exists():
-            ctx.fail(f"directory does not exist: {dirpath}")
+            raise OSError(f"directory does not exist: {dirpath}")
```
Comment (cdoern) [inline src/instructlab/data/list.py:38]: FileNotFoundError or is there a DirectoryNotFoundError?

**Review by cdoern — COMMENTED** [review]
nice, just a few comments

File: src/instructlab/cli/data/data.py (hunk 3 above)
Comment (jaideepr97) [reply src/instructlab/cli/data/data.py:23]: I haven't touched the code in this file, only moved it over

File: src/instructlab/data/list.py (hunk 4 above)
Comment (jaideepr97) [reply src/instructlab/data/list.py]: changed to `FileNotFoundError`, thanks

File: src/instructlab/cli/data/data.py (hunk 3 above)
Comment (nathan-weinberg) [reply src/instructlab/cli/data/data.py]: If that's the case let's do it as a followup - I'd like to change as little as possible following the large-size E2E success

**Review by nathan-weinberg — APPROVED** [review]
Pending CI passing this LGTM - thanks @jaideepr97!

File: pyproject.toml [hunk 5]
```diff
@@ -49,8 +49,8 @@ issues = "https://github.com/instructlab/instructlab/issues"
 "show" = "instructlab.config.show:show"
 
 [project.entry-points."instructlab.command.data"]
-"generate" = "instructlab.data.generate:generate"
-"list" = "instructlab.data.list:list_data"
+"generate" = "instructlab.cli.data.generate:generate"
```
Comment (alinaryan) [inline pyproject.toml:52]: does the alias section also need to be altered? line 73

Comment (jaideepr97) [reply pyproject.toml:52]: good catch! updated

**Review by danmcp — APPROVED** [review]

**Pull request merged**